pkg/controllers/util/federatedclient/podinformer.go

```diff
@@ -0,0 +1,109 @@
+package federatedclient
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/informers"
+	kubeclient "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+)
+
+func (f *federatedClientFactory) addPodInformer(informer informers.SharedInformerFactory, client kubeclient.Interface) {
+	informer.InformerFor(&corev1.Pod{}, func(k kubeclient.Interface, duration time.Duration) cache.SharedIndexInformer {
+		return cache.NewSharedIndexInformer(
+			f.podListerWatcher(client),
+			&corev1.Pod{},
+			0,
+			cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
+		)
+	})
+}
+
+func (f *federatedClientFactory) podListerWatcher(client kubeclient.Interface) cache.ListerWatcher {
+	return &cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			cur := atomic.LoadInt32(&f.availablePodListers)
+			if cur == 0 {
+				return nil, fmt.Errorf("do not have enought lister")
+			}
+			if cur > 0 {
+				if !atomic.CompareAndSwapInt32(&f.availablePodListers, cur, cur-1) {
+					return nil, fmt.Errorf("failed to compete for lister")
+				}
+				defer atomic.AddInt32(&f.availablePodListers, 1)
```

Comment:
So we immediately release the semaphore after this function returns, but the list is not GC'ed yet because it is yet to be consumed by the caller of this list function. Would this defeat the memory surge optimization we are trying to achieve?